step-functions/step-functions-client-api/src/main/java/step/functions/execution/FunctionExecutionService.java

```diff
@@ -39,6 +40,12 @@ public interface FunctionExecutionService {
 	
 	void returnTokenHandle(String tokenHandleId) throws FunctionExecutionServiceException;
 
-	<IN,OUT> Output<OUT> callFunction(String tokenHandleId, Function function, FunctionInput<IN> functionInput, Class<OUT> outputClass);
+	/**
+	 * Call function from while running the execution
+	 */
+	default <IN,OUT> Output<OUT> callFunction(String tokenHandleId, Function function, FunctionInput<IN> functionInput, Class<OUT> outputClass, AbstractStepContext executionContext){

```

Comment:
**To clarify:** Insteadof changing the signature of this method I'm wondering if it wouldn't be easier and cleaner to create a copy of the FunctionExecutionServiceImpl for each execution and store it to the context like here: step.engine.plugins.FunctionPlugin#initializeExecutionEngineContext

Reply:
@jeromecomte 
I'll think how to improve this. The main reason why I changed the signature is that the `registerFile()` is finally called from the functionType object (like `JMeterFunctionType`). 

But each FunctionType in system is registered in `functionTypeRegistry` via `@Plugin` (for instance, `JMeterPlugin#serverStart()`).

So `functionType` are singletones, and it is hard to bind them to current execution context without adding the context to `AbstractFunctionType#getHandlerProperties()`. That is, the signature of `AbstractFunctionType#getHandlerProperties()` will be changed anyway...